File: src/c_uuid_v7/_core/_windows.c
#ifdef _WIN32

#include "_platform.h"

#include <bcrypt.h>
#pragma comment(lib, "bcrypt.lib")

uint64_t epoch_base_ms = 0;
uint64_t tick_base_ms = 0;
#if PY_VERSION_HEX < 0x030E0000
VOID(WINAPI *query_interrupt_time_ptr)(PULONGLONG) = NULL;
#endif

uint64_t system_ms(void) {
    FILETIME ft;
    ULARGE_INTEGER ticks;

    GetSystemTimePreciseAsFileTime(&ft);
    ticks.QuadPart = (uint64_t)ft.dwHighDateTime << 32 | (uint64_t)ft.dwLowDateTime;
    return (ticks.QuadPart - 116444736000000000ULL) / 10000ULL;
}

int fill_random(unsigned char *buf, const Py_ssize_t len) {
    const NTSTATUS status = BCryptGenRandom(NULL, buf, (ULONG)len, BCRYPT_USE_SYSTEM_PREFERRED_RNG);
    return status >= 0 ? 0 : -1;
}

void platform_seeded(void) {
    epoch_base_ms = system_ms();
#if PY_VERSION_HEX >= 0x030E0000
    /* Python 3.14+ requires Windows 10+, where QueryInterruptTime is available. */
    {
        ULONGLONG interrupt_time = 0;

        QueryInterruptTime(&interrupt_time);
        tick_base_ms = interrupt_time / 10000ULL;
    }
#else
    /* Python 3.10-3.13 still support Windows 8.1, so QueryInterruptTime may be absent. */
    HMODULE kernel32 = GetModuleHandleA("kernel32.dll");

    query_interrupt_time_ptr = NULL;
    if (kernel32 != NULL) {
        query_interrupt_time_ptr =
            (VOID(WINAPI *)(PULONGLONG))GetProcAddress(kernel32, "QueryInterruptTime");
    }

    if (query_interrupt_time_ptr != NULL) {
        ULONGLONG interrupt_time = 0;

        query_interrupt_time_ptr(&interrupt_time);
        tick_base_ms = interrupt_time / 10000ULL;
        return;
    }

    tick_base_ms = GetTickCount64();
#endif
}

#endif

File: src/windows.zig
const builtin = @import("builtin");
const state = @import("state.zig");

const c = @import("c.zig").c;

const win = @cImport({
    @cDefine("WIN32_LEAN_AND_MEAN", "1");
    @cInclude("windows.h");
    @cInclude("bcrypt.h");
});

fn ensureBcryptGenRandom() ?state.BCryptGenRandomFn {
    if (builtin.os.tag != .windows) {
        return null;
    }

    if (state.runtime.bcrypt_gen_random_ptr) |bcrypt_gen_random| {
        return bcrypt_gen_random;
    }

    const bcrypt_module = win.GetModuleHandleA("bcrypt.dll") orelse win.LoadLibraryA("bcrypt.dll");
    if (bcrypt_module == null) {
        return null;
    }

    state.runtime.bcrypt_gen_random_ptr = @ptrCast(win.GetProcAddress(bcrypt_module, "BCryptGenRandom"));
    return state.runtime.bcrypt_gen_random_ptr;
}

pub fn nowMs() u64 {
    if (state.runtime.query_interrupt_time_ptr) |query_interrupt_time| {
        var interrupt_time: u64 = 0;
        query_interrupt_time(&interrupt_time);
        return state.runtime.epoch_base_ms + @as(u64, interrupt_time / 10_000) - state.runtime.tick_base_ms;
    }

    return state.runtime.epoch_base_ms + @as(u64, win.GetTickCount64()) - state.runtime.tick_base_ms;
}

pub fn systemMs() u64 {
    var ft: win.FILETIME = undefined;

    win.GetSystemTimePreciseAsFileTime(&ft);
    const ticks = (@as(u64, ft.dwHighDateTime) << 32) | @as(u64, ft.dwLowDateTime);
    return (ticks - 116_444_736_000_000_000) / 10_000;
}

pub fn fillRandom(buf: [*]u8, len: isize) c_int {
    const bcrypt_gen_random = ensureBcryptGenRandom() orelse return -1;
    const status = bcrypt_gen_random(null, buf, @intCast(len), win.BCRYPT_USE_SYSTEM_PREFERRED_RNG);
    return if (status >= 0) 0 else -1;
}

pub fn platformSeeded() void {
    const kernel32 = win.GetModuleHandleA("kernel32.dll");

    state.runtime.epoch_base_ms = systemMs();
    state.runtime.query_interrupt_time_ptr = null;

    if (kernel32 != null) {
        state.runtime.query_interrupt_time_ptr = @ptrCast(win.GetProcAddress(kernel32, "QueryInterruptTime"));
    }

    if (state.runtime.query_interrupt_time_ptr) |query_interrupt_time| {
        var interrupt_time: u64 = 0;

        query_interrupt_time(&interrupt_time);
        state.runtime.tick_base_ms = @as(u64, interrupt_time / 10_000);
        return;
    }

    state.runtime.tick_base_ms = @as(u64, win.GetTickCount64());
}
